### PR TITLE
Pull latest changes (incl. 2.2.1)

### PR DESCRIPTION
**Pull request by jibsaram**

### CODE REVIEW

**Comment by jibsaram** [comment]
Yikes! This is completely the opposite way from my intention 🤦‍♂️ . Apologies! I can never figure out how Github's arrows should be read, apparently. Terribly sorry about this!

**Comment by Pathoschild** [comment]
No worries! There was a history rewrite because SMAPI 2.11.1/3.0 were in parallel branches until 2.11.1 released, which is why the merge was needed. Looks like your fork doesn't have changes yet, so I suggest resetting to `develop` instead of merging (or just deleting the fork and reforking).

**Comment by jibsaram** [comment]
Yeah I was being silly, I realized my local repository was actually clone of your main repo instead of my fork, so I wanted to fix that plus pull your recent-most changes into mine. I think it's all fine now, although my heart might need another moment to calm down after I attempted to merge _err, something_  into your actual repository. Yikes!

Now that I have you anyway, I just pushed what I've been working on today to a separate branch (I'll get the hang it) over at my fork. It basically introduces player event listening to all players connected, not just the main player. I am certain there's many bugs, flaws and other issues, but I needed this for a project I'm working on and was hoping that this might be of some use, in case the main SMAPI project can benefit from this functionality too (and, of course, if my code is somewhat salvageable — I am extremely new to C#, I must admit).

I originally started by creating separate FarmerEvents alongside PlayerEvents, but realized it would be cleaner to just have the one, and since the actual farmer dude(tte) is called `farmer` I renamed the whole shebang. Not very backwards compatible, but should this end up being something you'd like to make a part of the main code, it might be a change that can be included with the upcoming big 3.0 update, as that deprecates a few other things. That, or a revert to just calling it all `Player*` of course.

Long story short; if you have a moment and [want to take a peek](https://github.com/hellodeibu/SMAPI/tree/farmerevents), I would love it if this might be of use to others as-well!

Cheers!

**Comment by Pathoschild** [comment]
Thanks for the work!

Yep, I'd rather keep the current event names. `PlayerEvents` refers to the concept (not the type name), and it's too late for new breaking changes in 3.0 so it'd be deprecated until 4.0 next year.

There's a complication though. If the current player isn't hosting, the `Warped` event will be raised correctly for other players but the event data will be completely wrong because it's not synced (unless they happen to warp into an [active location](https://stardewvalleywiki.com/Modding:Modder_Guide/APIs/Multiplayer#Get_active_locations)). That's the main reason it hasn't been implemented for other players yet.

**Comment by jibsaram** [comment]
I see your point, and it makes sense. I'll update my code to bring `PlayerEvents` back, that should fix a large part of what I have broken with these changes. 

As for the other part, I hadn't thought of the scenario when someone isn't the host! I think at least for my use-case I'll be able to add a check to ensure events for other players are —for now, at least— only triggered when the person is the host of the game, but I see how that would be less ideal from a mod creator point of view (you lose some predictability here). There's certainly room for improvement!

I'll at least apply these two changes so that things will at least not break, and perhaps we can revisit this feature later on to see if there's a good enough way to merge things down the road. 

Thanks so much for checking and replying!
